oldrunner.6
.\" 	$Id: oldrunner.6,v 1.2 2012/01/06 10:13:55 culot Exp $
.\"
.\" Copyright (c) 2010, 2012 Frederic Culot <frederic@culot.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"      - Redistributions of source code must retain the above
.\"        copyright notice, this list of conditions and the
.\"        following disclaimer.
.\"
.\"      - Redistributions in binary form must reproduce the above
.\"        copyright notice, this list of conditions and the
.\"        following disclaimer in the documentation and/or other
.\"        materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.fam C
.TH OLDRUNNER 6 "January 31, 2012" "OLDRUNNER(6)" "Oldrunner Manual"
.SH NAME
Oldrunner \- Broderbund's Loderunner remake
.SH SYNOPSIS
.B "oldrunner "
[\fB-v\fP] [\fB-l\fP<\fInum\fP>]
.SH DESCRIPTION
Oldrunner is a remake of Broderbund's Loderunner which contains the
150 original game levels. In order to win a level, the player '\fB@\fP'
must collect all gold bags '\fB$\fP' without being touched by ennemies
(same sprite '\fB@\fP' as player but in reverse video). Once all gold
is collected an escape ladder usually appears and the player must
travel to the top of the screen to reach the next level. The player can
dig holes into floors to temporarily trap ennemies and may then safely
walk atop them. Over time, floors dug into will regenerate, filling in
these holes. A trapped ennemy who cannot escape a hole before it fills
is consumed, immediately respawning in a random location at the top of
the level.
.SH OPTIONS
The following options are supported:
.TP
\fB-l\fP <num>
Start the game at level number \fInum\fP.
.TP
\fB-v\fP
Display oldrunner version and exit.
.SH CONFIGURATION
The game can be configured by modifying the variables found in
the cfg.h header file and recompiling. The default keys are the
following (\fB^X\fP means \fBcontrol\fP and \fBX\fP keys pressed
simultaneously):
.TP
\fBh\fP
move left
.TP
\fBl\fP
move right
.TP
\fBk\fP
move up
.TP
\fBj\fP
move down
.TP
\fBs\fP
dig on the left
.TP
\fBd\fP
dig on the right
.TP
\fB^g\fP
commit suicide
.TP
\fB^n\fP
go to next level
.TP
\fB^p\fP
go back to previous level
.TP
\fB^d\fP
quit the game
.SH SEE ALSO
ncurses(3)
.br
oldrunner web page: http://culot.org/public/Code/oldrunner.html
.SH AUTHOR
\fBFrederic Culot\fP <frederic@culot.org>.
.SH COPYRIGHT
Copyright (c) 2010, 2012 by Frederic Culot. 
.br
This software is released under the BSD License.


ChangeLog
2012-01-31  Frederic Culot  <frederic@culot.org>

	* Fix a display issue which caused part of the game board to be
	hidden on wide screens.

	* Fix an erroneous check which could lead to an out of bounds
	access on the curlvl->lay array.

2012-01-06  Frederic Culot  <frederic@culot.org>

	* Fix the sprite enum to avoid indexes past the end of arrays such
	as with the SP_INVALID element in sprites_init().

2010-07-17  Frederic Culot  <frederic@culot.org>

	* first public release

2010-04-01  Frederic Culot  <frederic@culot.org>

	* beginning of the project


oldrunner.c
/*  $Id: oldrunner.c,v 1.1.1.1 2010/07/17 17:30:32 culot Exp $  */

/*
 * Copyright (c) 2010 Frederic Culot <frederic@culot.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the
 *        following disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the
 *        following disclaimer in the documentation and/or other
 *        materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <limits.h>
#include <stdio.h>
#include <unistd.h>

#include "oldrunner.h"

/*
 * Oldrunner is a remake of the Loadrunner game from Broderbund, and one of the
 * first computer games my father offered me. I wrote this in memory of him.
 *
 *      Jean-Francois Culot
 *    1952-Sep-30 - 2009-Oct-20
 */
int
main (int argc, char **argv)
{
  const char *errstr;
  int ch, startlvl;

  startlvl = 0;
  while ((ch = getopt (argc, argv, "l:v")) != -1)
    {
      switch (ch)
        {
        case 'l':
          startlvl = (int)strtonum (optarg, 0, INT_MAX, &errstr) - 1;
          if (errstr)
            {
              fprintf (stderr, "Bad level number: %s\n", errstr);
              return 1;
            }
          break;
        case 'v':
          fputs ("oldrunner-" VERSION "\n", stdout);
          return 0;
        default:
          goto err;
        }
    }
  argc -= optind;
  argv += optind;
  if (argc >= 1)
    {
    err:
      fputs ("usage: oldrunner [-v] [-l<num>]\n", stderr);
      return 0;
    }

  game_init (startlvl);
  usr_input ();
  game_end ();
  
  return 0;
}
